# Project 3: Build Decision Tree for Bank Loan Modelling
DV - "Personal Loan"
IDV - Output of RF Algorithm

In [1]:
import pandas as pd
import numpy as np
from sklearn import tree
from sklearn import preprocessing
from warnings import filterwarnings
from sklearn.ensemble import RandomForestClassifier
filterwarnings('ignore')

In [2]:
dataset=pd.read_excel('Bank_Personal_Loan_Modelling.xlsx',sheet_name=1)
dataset.head()

ID  Age  Experience  Income  ZIP Code  Family  CCAvg  Education  Mortgage  \
0   1   25           1      49     91107       4    1.6          1         0   
1   2   45          19      34     90089       3    1.5          1         0   
2   3   39          15      11     94720       1    1.0          1         0   
3   4   35           9     100     94112       1    2.7          2         0   
4   5   35           8      45     91330       4    1.0          2         0   

   Personal Loan  Securities Account  CD Account  Online  CreditCard  
0              0                   1           0       0           0  
1              0                   1           0       0           0  
2              0                   0           0       0           0  
3              0                   0           0       0           0  
4              0                   0           0       0           1

In [3]:
dataset.shape

(5000, 14)

In [4]:
dataset.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5000 entries, 0 to 4999
Data columns (total 14 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   ID                  5000 non-null   int64  
 1   Age                 5000 non-null   int64  
 2   Experience          5000 non-null   int64  
 3   Income              5000 non-null   int64  
 4   ZIP Code            5000 non-null   int64  
 5   Family              5000 non-null   int64  
 6   CCAvg               5000 non-null   float64
 7   Education           5000 non-null   int64  
 8   Mortgage            5000 non-null   int64  
 9   Personal Loan       5000 non-null   int64  
 10  Securities Account  5000 non-null   int64  
 11  CD Account          5000 non-null   int64  
 12  Online              5000 non-null   int64  
 13  CreditCard          5000 non-null   int64  
dtypes: float64(1), int64(13)
memory usage: 547.0 KB


In [5]:
# Checking null values
dataset.isnull().sum()

ID                    0
Age                   0
Experience            0
Income                0
ZIP Code              0
Family                0
CCAvg                 0
Education             0
Mortgage              0
Personal Loan         0
Securities Account    0
CD Account            0
Online                0
CreditCard            0
dtype: int64

In [6]:
# Dropping unnecessary columns
dataset=dataset.drop(['ID','ZIP Code'],axis=1)
dataset.head()

Age  Experience  Income  Family  CCAvg  Education  Mortgage  Personal Loan  \
0   25           1      49       4    1.6          1         0              0   
1   45          19      34       3    1.5          1         0              0   
2   39          15      11       1    1.0          1         0              0   
3   35           9     100       1    2.7          2         0              0   
4   35           8      45       4    1.0          2         0              0   

   Securities Account  CD Account  Online  CreditCard  
0                   1           0       0           0  
1                   1           0       0           0  
2                   0           0       0           0  
3                   0           0       0           0  
4                   0           0       0           1

In [7]:
# Splitting the Dataset
x=dataset.drop(['Personal Loan'],axis=1)
x.head()

Age  Experience  Income  Family  CCAvg  Education  Mortgage  \
0   25           1      49       4    1.6          1         0   
1   45          19      34       3    1.5          1         0   
2   39          15      11       1    1.0          1         0   
3   35           9     100       1    2.7          2         0   
4   35           8      45       4    1.0          2         0   

   Securities Account  CD Account  Online  CreditCard  
0                   1           0       0           0  
1                   1           0       0           0  
2                   0           0       0           0  
3                   0           0       0           0  
4                   0           0       0           1

In [8]:
y=dataset.iloc[:,7:8]
y.head()

Personal Loan
0              0
1              0
2              0
3              0
4              0

In [9]:
x.shape

(5000, 11)

In [10]:
y.shape

(5000, 1)

In [11]:
# Building and fitting the model
tree_model=tree.DecisionTreeClassifier()
tree_model.fit(X=x,y=y)

DecisionTreeClassifier(ccp_alpha=0.0, class_weight=None, criterion='gini',
                       max_depth=None, max_features=None, max_leaf_nodes=None,
                       min_impurity_decrease=0.0, min_impurity_split=None,
                       min_samples_leaf=1, min_samples_split=2,
                       min_weight_fraction_leaf=0.0, presort='deprecated',
                       random_state=None, splitter='best')

In [12]:
# Checking the model score/accuracy for prediction
tree_model.score(X=x,y=y)

1.0

The Model is 100% Accurate for prediction

In [13]:
# Now Applying Random Forest to identify the Important IDV for Decision Tree
rf_model=RandomForestClassifier(n_estimators=1000,max_features=2,oob_score=True)

In [14]:
# fitting the RF model
rf_model.fit(X=x,y=y)

RandomForestClassifier(bootstrap=True, ccp_alpha=0.0, class_weight=None,
                       criterion='gini', max_depth=None, max_features=2,
                       max_leaf_nodes=None, max_samples=None,
                       min_impurity_decrease=0.0, min_impurity_split=None,
                       min_samples_leaf=1, min_samples_split=2,
                       min_weight_fraction_leaf=0.0, n_estimators=1000,
                       n_jobs=None, oob_score=True, random_state=None,
                       verbose=0, warm_start=False)

In [15]:
# Checking the Accuracy Score
print("OOB Accuracy Score:",rf_model.oob_score_)

OOB Accuracy Score: 0.9874


In [16]:
# Checking the feature importances
for feature,imp in zip(x,rf_model.feature_importances_):
    print(feature,":",imp)

Age : 0.043181741524203704
Experience : 0.04409188356634103
Income : 0.3500689478335168
Family : 0.09924600719183353
CCAvg : 0.17860746534551358
Education : 0.1651231853697049
Mortgage : 0.04400432601823044
Securities Account : 0.00560220832116716
CD Account : 0.05217643735035223
Online : 0.008271611599231813
CreditCard : 0.009626185879904825


Here we observe that the following features are having their values >=0.05.
1. Income
2. Family
3. CCAvg
4. Education
5. CD Account

These all are the important features which affect the Personal Loan.

In [17]:
# Splitting the IDV i.e Important features
imp_features=dataset[['Income','Family','CCAvg','Education','CD Account']]
imp_features.head()

Income  Family  CCAvg  Education  CD Account
0      49       4    1.6          1           0
1      34       3    1.5          1           0
2      11       1    1.0          1           0
3     100       1    2.7          2           0
4      45       4    1.0          2           0

In [18]:
# Building the Decision Tree Model (Here depth=10 means we have 2 outputs Personal Loan 1 and Personal Loan 0 and we have 5 IDV i.e 2*5=10)
tree_model=tree.DecisionTreeClassifier(max_depth=10)

In [19]:
# Fitting the Decision Tree Model
tree_model.fit(X=imp_features,y=y)

DecisionTreeClassifier(ccp_alpha=0.0, class_weight=None, criterion='gini',
                       max_depth=10, max_features=None, max_leaf_nodes=None,
                       min_impurity_decrease=0.0, min_impurity_split=None,
                       min_samples_leaf=1, min_samples_split=2,
                       min_weight_fraction_leaf=0.0, presort='deprecated',
                       random_state=None, splitter='best')

In [20]:
# Creating the Decision Tree
with open("Dtree.dot","w") as f:
    f=tree.export_graphviz(tree_model,feature_names=["Income","Family","CCAvg","Education","CD Account"],out_file=f);

In [21]:
# Checking the model score/accuracy for prediction
tree_model.score(X=imp_features,y=y)

0.9972

The Model is 99.7% Accurate for prediction

In [22]:
# Fitting the RF model
rf_model.fit(X=imp_features,y=y)

RandomForestClassifier(bootstrap=True, ccp_alpha=0.0, class_weight=None,
                       criterion='gini', max_depth=None, max_features=2,
                       max_leaf_nodes=None, max_samples=None,
                       min_impurity_decrease=0.0, min_impurity_split=None,
                       min_samples_leaf=1, min_samples_split=2,
                       min_weight_fraction_leaf=0.0, n_estimators=1000,
                       n_jobs=None, oob_score=True, random_state=None,
                       verbose=0, warm_start=False)

In [23]:
# Checking the Accuracy Score
print("OOB Accuracy Score:",rf_model.oob_score_)

OOB Accuracy Score: 0.989


The OOB Accuracy is 98.9% which is almost considered as 99%

# Rules of Decision Tree
1) 3629 Customers having income less than 106.5 and CCavg less than 2.95 did not get loan.

2) 314 Customers having income greater than 116.5 with education level of either Graduate or Advanced/Professional got the loan.

3) 69 Customers having income greater than 113.5 but less than 116.5 with education level of Undergraduate and family size greater than 2.5 got the loan.

4) 566 Customers having income greater than 113.5 but less than 116.5 with education level of Undergraduate and family size less than 2.5 did not get the loan.

5) 11 Customers having income less than 106.5 with CCAvg greater than 2.95 but less than 3.85 and CD Account greater than 0.5 got the loan.  